src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs

```diff
@@ -79,90 +96,113 @@ public class PipelineArtifactServer
+                    BuildArtifact buildArtifact;
+                    if (downloadParameters.ProjectRetrievalOptions == BuildArtifactRetrievalOptions.RetrieveByProjectId)
                     {
-                        throw new InvalidOperationException("Project name can't be empty when trying to fetch build artifacts!");
+                        buildArtifact = await buildHelper.GetArtifact(downloadParameters.ProjectId, downloadParameters.PipelineId, downloadParameters.ArtifactName, cancellationToken);
+                    }
+                    else if (downloadParameters.ProjectRetrievalOptions == BuildArtifactRetrievalOptions.RetrieveByProjectName)
+                    {
+                        if (string.IsNullOrEmpty(downloadParameters.ProjectName))
+                        {
+                            throw new InvalidOperationException("Project name can't be empty when trying to fetch build artifacts!");
+                        }
+                        else
+                        {
+                            buildArtifact = await buildHelper.GetArtifactWithProjectNameAsync(downloadParameters.ProjectName, downloadParameters.PipelineId, downloadParameters.ArtifactName, cancellationToken);
+                        }
                     }
                     else
                     {
-                        buildArtifact = await buildHelper.GetArtifactWithProjectNameAsync(downloadParameters.ProjectName, downloadParameters.PipelineId, downloadParameters.ArtifactName, cancellationToken);
+                        throw new InvalidOperationException("Unreachable code!");
                     }
+
+                    var manifestId = DedupIdentifier.Create(buildArtifact.Resource.Data);
+                    var options = DownloadPipelineArtifactOptions.CreateWithManifestId(
+                        manifestId,
+                        downloadParameters.TargetDirectory,
+                        proxyUri: null,
+                        minimatchPatterns: downloadParameters.MinimatchFilters);
+                    
+                    PipelineArtifactActionRecord downloadRecord = clientTelemetry.CreateRecord<PipelineArtifactActionRecord>((level, uri, type) =>
+                            new PipelineArtifactActionRecord(level, uri, type, nameof(DownloadAsync), context));
+                    await clientTelemetry.MeasureActionAsync(
+                        record: downloadRecord,
+                        actionAsync: async () =>
+                        {
+                            await dedupManifestClient.DownloadAsync(options, cancellationToken);                            
+                        });
+                    // Send results to CustomerIntelligence
+                    context.PublishTelemetry(area: "AzurePipelinesAgent", feature: "PipelineArtifact", record: downloadRecord);  
                 }
                 else
                 {
                     throw new InvalidOperationException("Unreachable code!");
```

Comment:
>Unreachable code! [](start = 57, length = 17)

Should be more helpful that unreachable code, call out something like invalid download option specified